### ISSUE TITLE
[JENKINS-56390] Duplicate ChangeLog on Jenkins builds for concurrent builds

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3409]
<p>The Git ChangeLog's are repeated in the builds as shown below.</p>

<p>This is pipeline job and the build takes one hour to complete and hence below 3 build executed in parallel but i see the changelog is repeated.</p>

<p>Versions used:</p>

<p><a href="https://jenkins.io/" class="external-link" target="_blank" rel="nofollow noopener">Jenkins ver. 2.121.3</a></p>

<p>Git Changelog plugin version  2.16</p>

<p>git-client 2.7.0</p>

<p>git 3.7.1</p>

<p> </p>

<p><b>ChangeLog</b></p>

<p><b>#12 (Jan 29, 2019 8:59:07 AM)</b><br/>
JIRA-XXX: Added validation — User3<br/>
<span class="error">&#91;JIRA-XXX&#93;</span> clean up — User1<br/>
JIRA-XXX: Cleanup — User2<br/>
JIRA-XXX: State model and final cleanup - user1</p>

<p><b>#10 (Jan 29, 2019 8:54:55 AM)</b><br/>
JIRA-XXX: Added validation — User3<br/>
<span class="error">&#91;JIRA-XXX&#93;</span> clean up — User1<br/>
JIRA-XXX: Cleanup — User2</p>

<p><b>#9 (Jan 29, 2019 8:45:43 AM)</b><br/>
JIRA-XXX: Added validation — User3</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hariprasadnr9">hariprasadnr9</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-56390" target="_blank">Duplicate ChangeLog on Jenkins builds for concurrent builds</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Critical
<li><b>component(s)</b>: git-plugin, workflow-scm-step-plugin
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2019-03-09T14:17:42+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>The Git ChangeLog's are repeated in the builds as shown below.</p>

<p>This is pipeline job and the build takes one hour to complete and hence below 3 build executed in parallel but i see the changelog is repeated.</p>

<p>Versions used:</p>

<p><a href="https://jenkins.io/" class="external-link" target="_blank" rel="nofollow noopener">Jenkins ver. 2.121.3</a></p>

<p>Git Changelog plugin version  2.16</p>

<p>git-client 2.7.0</p>

<p>git 3.7.1</p>

<p> </p>

<p><b>ChangeLog</b></p>

<p><b>#12 (Jan 29, 2019 8:59:07 AM)</b>
JIRA-XXX: Added validation — User3
<span class="error">&#91;JIRA-XXX&#93;</span> clean up — User1
JIRA-XXX: Cleanup — User2
JIRA-XXX: State model and final cleanup - user1</p>

<p><b>#10 (Jan 29, 2019 8:54:55 AM)</b>
JIRA-XXX: Added validation — User3
<span class="error">&#91;JIRA-XXX&#93;</span> clean up — User1
JIRA-XXX: Cleanup — User2</p>

<p><b>#9 (Jan 29, 2019 8:45:43 AM)</b>
JIRA-XXX: Added validation — User3</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-56390?focusedId=362012&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362012">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The changelog wants to present the changes since the last completed build.  Since the 3 jobs you listed all ran in parallel and had not completed. the changes list may be correct displaying it for each of the jobs.</p>
</pre>
</details>

</ul>
</details>

<p>The changelog wants to present the changes since the last completed build.  Since the 3 jobs you listed all ran in parallel and had not completed. the changes list may be correct displaying it for each of the jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hariprasadnr9:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-56390?focusedId=362091&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362091">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a>,</p>

<p>Yes i understand that. Even i found out that the root cause of this is what is you mentioned.</p>

<p> </p>

<p>But in case the build fails i am fetching the emails from the change log and all the developer who did commits are notified.</p>

<p>Instead only one developer who did the wrong commit at the end has to be notified.</p>

<p> </p>

<p>This is where i started to analyze why it notified also other developers which commited previous to the failing commit.</p>

<p>Also for the developers its confusing when they look at change log since there are multiple commits.</p>

<p>Hence displaying only the change from previous git revision/ previously triggered build makes sense instead of last completed build.</p>

<p> </p>

<p>Please let me know if there is any resolution to fix this issue.</p>

<p> </p>

<p>Thanks,</p>

<p>Hariprasad</p>
</pre>
</details>

</ul>
</details>

<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a>,</p>

<p>Yes i understand that. Even i found out that the root cause of this is what is you mentioned.</p>

<p> </p>

<p>But in case the build fails i am fetching the emails from the change log and all the developer who did commits are notified.</p>

<p>Instead only one developer who did the wrong commit at the end has to be notified.</p>

<p> </p>

<p>This is where i started to analyze why it notified also other developers which commited previous to the failing commit.</p>

<p>Also for the developers its confusing when they look at change log since there are multiple commits.</p>

<p>Hence displaying only the change from previous git revision/ previously triggered build makes sense instead of last completed build.</p>

<p> </p>

<p>Please let me know if there is any resolution to fix this issue.</p>

<p> </p>

<p>Thanks,</p>

<p>Hariprasad</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hariprasadnr9:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-56390?focusedId=362106&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362106">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ci_jenkinsci_org" class="user-hover" rel="ci_jenkinsci_org">ci_jenkinsci_org</a></p>

<p>Is it possible to fix this issue using Git ChangeLog Plugin or Last Changes plugin in Jenkins.</p>

<p> </p>

<p>Thanks,</p>

<p>Hariprasad</p>
</pre>
</details>

</ul>
</details>

<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ci_jenkinsci_org" class="user-hover" rel="ci_jenkinsci_org">ci_jenkinsci_org</a></p>

<p>Is it possible to fix this issue using Git ChangeLog Plugin or Last Changes plugin in Jenkins.</p>

<p> </p>

<p>Thanks,</p>

<p>Hariprasad</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-56390?focusedId=362147&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362147">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hariprasadnr9" class="user-hover" rel="hariprasadnr9">hariprasadnr9</a> it is possible to avoid the issue by using pull requests rather than having everyone commit to a single branch.  A multibranch Pipeline could then be configured to create and delete jobs each time a developer creates a pull request to the central repository.  Refer to <a href="https://standard.cbu.cloudbees.com/cloudbees-university-jenkins-pipeline-fundamentals" class="external-link" target="_blank" rel="nofollow noopener">Jenkins Pipeline Fundamentals</a> for a free self-paced Jenkins Pipeline training course.</p>

<p>it is possible to fix the issue if you and your developers are willing to accept specific constraints.  If you prevent parallel builds and if you immediately run a build each time a change is pushed to the central repository, then the attribution of build breaks will be absolutely clear.  Most teams are unwilling to accept that delay.</p>

<p>Another fix is to reduce the build time so that it takes much less time to complete a build.  That decreases the time window where parallel jobs could be running and reduces the time from commit to completed build.</p>

<p>There isn't any code in any of the plugins that I know which will detect that a parallel build has been started with some of the commits being already included in a running parallel build.  It seems like the graph analysis to detect the condition is possible though I'm not sure how a user interface would be presented to the user to show them that earlier parallel builds are evaluating a subset of the changes   Jenkins multibranch Pipeline is the best of the alternatives that I've listed.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hariprasadnr9" class="user-hover" rel="hariprasadnr9">hariprasadnr9</a> it is possible to avoid the issue by using pull requests rather than having everyone commit to a single branch.  A multibranch Pipeline could then be configured to create and delete jobs each time a developer creates a pull request to the central repository.  Refer to <a href="https://standard.cbu.cloudbees.com/cloudbees-university-jenkins-pipeline-fundamentals" class="external-link" target="_blank" rel="nofollow noopener">Jenkins Pipeline Fundamentals</a> for a free self-paced Jenkins Pipeline training course.</p>

<p>it is possible to fix the issue if you and your developers are willing to accept specific constraints.  If you prevent parallel builds and if you immediately run a build each time a change is pushed to the central repository, then the attribution of build breaks will be absolutely clear.  Most teams are unwilling to accept that delay.</p>

<p>Another fix is to reduce the build time so that it takes much less time to complete a build.  That decreases the time window where parallel jobs could be running and reduces the time from commit to completed build.</p>

<p>There isn't any code in any of the plugins that I know which will detect that a parallel build has been started with some of the commits being already included in a running parallel build.  It seems like the graph analysis to detect the condition is possible though I'm not sure how a user interface would be presented to the user to show them that earlier parallel builds are evaluating a subset of the changes   Jenkins multibranch Pipeline is the best of the alternatives that I've listed.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>hariprasadnr9:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-56390?focusedId=362250&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362250">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a></p>

<p>Thank for the quick response.</p>

<p>I see all the above solutions as a work around.</p>

<p>We have the Jenkins Enterprise setup which can handle multiple builds to run in parallel.</p>

<p>What would be the solution to fix this issue.?</p>

<p>&gt; A plugin should be implemented when extended git functionality</p>

<p>&gt; Jenkins need to implement this feature to handle the changeLog when build are running in parallle</p>

<p> </p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hello <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a></p>

<p>Thank for the quick response.</p>

<p>I see all the above solutions as a work around.</p>

<p>We have the Jenkins Enterprise setup which can handle multiple builds to run in parallel.</p>

<p>What would be the solution to fix this issue.?</p>

<p>> A plugin should be implemented when extended git functionality</p>

<p>> Jenkins need to implement this feature to handle the changeLog when build are running in parallle</p>

<p> </p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-56390?focusedId=362324&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362324">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hariprasadnr9" class="user-hover" rel="hariprasadnr9">hariprasadnr9</a> I think multibranch pipelines are the best solution to the problem you're trying to solve.</p>

<p>Each developer wants to see the results of their changes as applied to the master branch.  They don't want the disruption of seeing the untested work of others intermixed with their changes.  A single job and a single branch means that the developers need to decide which of the builds within the job are theirs and they need to decide which of the commits are theirs within that build.  </p>

<p>If instead each developer pushed their changes that needed to be evaluated to a separate branch for the change they were proposing, then the multibranch pipeline will automatically create a job dedicated to that branch.  The developer will see only their changes on the new job for their new branch.  They can add more commits, amend their commits, invite code review of their commits, and when those things are complete, they can merge that branch into the master branch.  The multibranch pipeline insulates them from the unverified changes of others and allows them to do a better job of preparing their changes to be merged into the master branch.</p>

<p>The multibranch pipeline works well with plain git (using branches), with Bitbucket (using branches, pull requests, and optionally forked repositories), with Gitea (using branches, pull requests, and optionally forked repositories), and with GitHub (using branches, pull requests, and optionally forked repositories).  Please reconsider the capability of multibranch pipeline as a better solution to the problem you're trying to solve.  It is something you can implement immediately and integrates well with Blue Ocean and other components.</p>

<p>The free <a href="https://standard.cbu.cloudbees.com/cloudbees-university-jenkins-pipeline-fundamentals" class="external-link" target="_blank" rel="nofollow noopener">Jenkins Pipeline Fundamentals course</a> introduces the multibranch pipeline concept very well.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=hariprasadnr9" class="user-hover" rel="hariprasadnr9">hariprasadnr9</a> I think multibranch pipelines are the best solution to the problem you're trying to solve.</p>

<p>Each developer wants to see the results of their changes as applied to the master branch.  They don't want the disruption of seeing the untested work of others intermixed with their changes.  A single job and a single branch means that the developers need to decide which of the builds within the job are theirs and they need to decide which of the commits are theirs within that build.  </p>

<p>If instead each developer pushed their changes that needed to be evaluated to a separate branch for the change they were proposing, then the multibranch pipeline will automatically create a job dedicated to that branch.  The developer will see only their changes on the new job for their new branch.  They can add more commits, amend their commits, invite code review of their commits, and when those things are complete, they can merge that branch into the master branch.  The multibranch pipeline insulates them from the unverified changes of others and allows them to do a better job of preparing their changes to be merged into the master branch.</p>

<p>The multibranch pipeline works well with plain git (using branches), with Bitbucket (using branches, pull requests, and optionally forked repositories), with Gitea (using branches, pull requests, and optionally forked repositories), and with GitHub (using branches, pull requests, and optionally forked repositories).  Please reconsider the capability of multibranch pipeline as a better solution to the problem you're trying to solve.  It is something you can implement immediately and integrates well with Blue Ocean and other components.</p>

<p>The free <a href="https://standard.cbu.cloudbees.com/cloudbees-university-jenkins-pipeline-fundamentals" class="external-link" target="_blank" rel="nofollow noopener">Jenkins Pipeline Fundamentals course</a> introduces the multibranch pipeline concept very well.</p>
